iaso/tasks/process_mobile_bulk_upload.py

```diff
@@ -0,0 +1,163 @@
+from iaso.models import Project, Instance, InstanceFile
+from iaso.utils.s3_client import download_file
+
+INSTANCES_JSON = "instances.json"
+ORG_UNITS_JSON = "orgUnits.json"
+
+logger = logging.getLogger(__name__)
+
+
+@task_decorator(task_name="process_mobile_bulk_upload")
+def process_mobile_bulk_upload(user_id, project_id, zip_file_object_name, task=None):
+    start_date = datetime.now()
+    start_time = time.time()
+    the_task = task
+    the_task.report_progress_and_stop_if_killed(
+        progress_value=0,
+        progress_message=_("Starting"),
+        end_value=100,
+    )
+    user = User.objects.get(id=user_id)
+    project = Project.objects.get(id=project_id)
+
+    api_import = APIImport.objects.create(
+        user=user,
+        import_type="bulk",
+        json_body={"file": zip_file_object_name},
+    )
+
+    try:
+        logger.info(f"Downloading {zip_file_object_name} from S3...")
+        zip_file_path = download_file(zip_file_object_name)
+        logger.info("DONE.")
+
+        stats = {"new_org_units": 0, "new_instances": 0, "new_instance_files": 0}
+
+        with transaction.atomic():
+            with zipfile.ZipFile(zip_file_path, "r") as zip_ref:
+                if ORG_UNITS_JSON in zip_ref.namelist():
+                    with zip_ref.open(ORG_UNITS_JSON) as file:
+                        new_org_units = import_org_units(json.load(file), user, project.app_id)
+                        stats["new_org_units"] = len(new_org_units)
+                else:
+                    logger.info(f"The file {ORG_UNITS_JSON} does not exist in the zip file.")
+
+                if INSTANCES_JSON in zip_ref.namelist():
+                    with zip_ref.open(INSTANCES_JSON) as file:
+                        import_instances(json.load(file), user, project.app_id)
+                else:
+                    logger.info(f"The file {INSTANCES_JSON} does not exist in the zip file.")
+
+                logger.info("Processing forms and files")
+                new_instance_files = []
+                for directory in zipfile.Path(zip_ref).iterdir():
+                    if not directory.is_dir():
+                        continue
+
+                    instance = process_instance_xml(directory, zip_ref, user)
+                    stats["new_instances"] += 1
+                    new_instance_files += process_instance_files(directory, instance)
+
+                duplicated_count = duplicate_instance_files(new_instance_files)
+                stats["new_instance_files"] = len(new_instance_files) + duplicated_count
+
+    except Exception as e:
+        logger.error("Exception! Rolling back import: " + str(e))
+        api_import.has_problem = True
+        api_import.exception = format_exc()
+        api_import.save()
+        raise e
+
+    the_task.report_success_with_result(
+        message=result_message(user, project, start_date, start_time, stats),
+    )
+
+
+def process_instance_xml(directory, zip_ref, user):
```

Comment:
Isn't this duplicated code from /sync/views.py 

In this case, we should make it a function as to ensure that when we update the code somewhere with new features, we don't forget to update it everywhere

Reply:
:+1: agreed, I've extracted a function